# Manipulace s `DataFrames`

## Příprava prostředí

In [81]:
import pandas
import os
import random
import sqlalchemy as sql    # alias pro knihovnu

In [82]:
__file__ = __vsc_ipynb_file__
SCRIPT_FOLDER = os.path.dirname(__file__)
PROJECT_FOLDER = os.path.dirname(SCRIPT_FOLDER)
DATA_FOLDER = os.path.join(PROJECT_FOLDER,'data')
SOURCE_FOLDER = os.path.join(PROJECT_FOLDER,'source')

# Pandas čtení souborů

In [83]:
my_csv = pandas.read_csv(r"C:\Users\Dorota\Desktop\datová analytika\working\data\source\sklad\cenik_mobilnich_telefonu.csv")

In [84]:
type(my_csv)
my_csv.info()
print("sloupce v našem DF:",my_csv.columns)
print("Index v našem DF:",my_csv.index)
print("počet řádů, sloupců",my_csv.shape)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 118 entries, 0 to 117
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Model         118 non-null    object 
 1   Barva         118 non-null    object 
 2   5G            59 non-null     object 
 3   4G LTE        91 non-null     object 
 4   VoLTE/Wi-Fi   72 non-null     object 
 5   Dual SIM      76 non-null     object 
 6   eSIM          40 non-null     object 
 7   Cena bez DPH  118 non-null    float64
 8   Cena s DPH    118 non-null    int64  
dtypes: float64(1), int64(1), object(7)
memory usage: 8.4+ KB
sloupce v našem DF: Index(['Model', 'Barva', '5G', '4G LTE', 'VoLTE/Wi-Fi', 'Dual SIM', 'eSIM',
       'Cena bez DPH', 'Cena s DPH'],
      dtype='object')
Index v našem DF: RangeIndex(start=0, stop=118, step=1)
počet řádů, sloupců (118, 9)


In [85]:
my_csv

Model           Barva   5G 4G LTE  \
0                     TCL TAB 10 4G Dark Gray           černá  NaN      ✓   
1                Alcatel 1 2021 Volcano Black           černá  NaN      ✓   
2                 Alcatel 2019G Metallic Gray  metalická šedá  NaN    NaN   
3                          Alcatel 1068 Black           černá  NaN    NaN   
4    Apple iPhone 14 Pro Max 1 TB Space Black  vesmírně černý    ✓      ✓   
..                                        ...             ...  ...    ...   
113   Sony PlayStation 5 2022 + hra a ovladač            bílá  NaN    NaN   
114         Microsoft XBOX Series S + ovladač            bílá  NaN    NaN   
115                          Yealink SIP-W73P            šedá  NaN    NaN   
116                          Yealink SIP-T43U            šedá  NaN    NaN   
117                          Yealink SIP-T33G            šedá  NaN    NaN   

    VoLTE/Wi-Fi Dual SIM eSIM  Cena bez DPH  Cena s DPH  
0           NaN      NaN  NaN       3534.71        4277  
1             ✓        ✓  NaN       1303.31        1577  
2           NaN      NaN  NaN        724.79         877  
3           NaN        ✓  NaN        394.21         477  
4             ✓        ✓    ✓      44195.87       53477  
..          ...      ...  ...           ...         ...  
113         NaN      NaN  NaN      14050.41       17001  
114         NaN      NaN  NaN       7934.71        9601  
115         NaN      NaN  NaN       2460.33        2977  
116         NaN      NaN  NaN       2212.40        2677  
117         NaN      NaN  NaN       1303.31        1577  

[118 rows x 9 columns]

## Přejmenování sloupců

In [86]:
file_mobily = r"C:\Users\Dorota\Desktop\datová analytika\working\data\source\sklad\cenik_mobilnich_telefonu.csv"

mb_cfg = {
    #'nrows':10,
    'usecols':[0,7,8],
    'skiprows': 1,
    'header': None,
    'names':['model', 'cena_bez_DPH', 'cena_s_DPH'],
    #'skipfooter': 10
}

mobily = pandas.read_csv(file_mobily, **mb_cfg)

mobily.info()
mobily

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 118 entries, 0 to 117
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   model         118 non-null    object 
 1   cena_bez_DPH  118 non-null    float64
 2   cena_s_DPH    118 non-null    int64  
dtypes: float64(1), int64(1), object(1)
memory usage: 2.9+ KB


model  cena_bez_DPH  cena_s_DPH
0                     TCL TAB 10 4G Dark Gray       3534.71        4277
1                Alcatel 1 2021 Volcano Black       1303.31        1577
2                 Alcatel 2019G Metallic Gray        724.79         877
3                          Alcatel 1068 Black        394.21         477
4    Apple iPhone 14 Pro Max 1 TB Space Black      44195.87       53477
..                                        ...           ...         ...
113   Sony PlayStation 5 2022 + hra a ovladač      14050.41       17001
114         Microsoft XBOX Series S + ovladač       7934.71        9601
115                          Yealink SIP-W73P       2460.33        2977
116                          Yealink SIP-T43U       2212.40        2677
117                          Yealink SIP-T33G       1303.31        1577

[118 rows x 3 columns]

In [87]:
file_kosm = r"C:\Users\Dorota\Desktop\datová analytika\working\data\source\sklad\cenik_kosmetiky.csv"

kosm_cfg = dict (
    #nrows=10,
    thousands=' ',
    usecols=[0,1,2,3],
    skiprows= 1,
    header= None,
    names=['kod','model', 'cena_bez_DPH', 'cena_s_DPH'],
    index_col=[0]
    )

kosmetika = pandas.read_csv(file_kosm,**kosm_cfg)
kosmetika.info()
kosmetika

<class 'pandas.core.frame.DataFrame'>
Index: 689 entries, 19106 to 16090
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   model         689 non-null    object 
 1   cena_bez_DPH  689 non-null    float64
 2   cena_s_DPH    689 non-null    float64
dtypes: float64(2), object(1)
memory usage: 21.5+ KB


model  cena_bez_DPH  \
kod                                                                   
19106              ČISTÍCÍ OLEJ NA PLEŤ 3 v 1 - 50 ml        277.69   
1910        ČISTÍCÍ OLEJ NA PLEŤ 3 v 1 - vzorek 13 ml         65.29   
19376               NEUTRÁLNÍ HYDROFILNÍ OLEJ - 50 ml        171.90   
1937         NEUTRÁLNÍ HYDROFILNÍ OLEJ - vzorek 13 ml         38.84   
19503                PLEŤOVÝ OLEJ RŮŽE-JASMÍN - 10 ml        166.94   
...                                               ...           ...   
5014                                LÁHEV 1 l - plast         17.36   
KD090006              TATSOI TANGA, PSP – DÁREK - 6 g         24.35   
KD020005    ŘEPA SANGRIA BÝČÍ KREV, PSP – DÁREK - 5 g         24.35   
D13502              MANDARINKA ČERVENÁ - DÁREK - 5 ml         78.51   
16090     TYMIÁN b.s. linalool/geraniol, PSP - 1.5 ml        203.80   

          cena_s_DPH  
kod                   
19106          336.0  
1910            79.0  
19376          208.0  
1937            47.0  
19503          202.0  
...              ...  
5014            21.0  
KD090006        28.0  
KD020005        28.0  
D13502          95.0  
16090          246.6  

[689 rows x 3 columns]

## Čtení dat z Internetu

In [88]:
kurzy_url = "https://www.cnb.cz/cs/financni-trhy/devizovy-trh/kurzy-devizoveho-trhu/kurzy-devizoveho-trhu/vybrane.txt?od=01.10.2025&do=31.10.2025&mena=EUR&format=txt"

cnb_cfg = dict(
    skiprows = 1,
    sep = '|',
    decimal = ',',
    parse_dates = ['Datum'],
    date_format = '%d.%m.%Y',
)

kurzy = pandas.read_table(kurzy_url,**cnb_cfg)

kurzy.info()
kurzy

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19 entries, 0 to 18
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   Datum   19 non-null     datetime64[ns]
 1   Kurz    19 non-null     float64       
dtypes: datetime64[ns](1), float64(1)
memory usage: 436.0 bytes


Datum    Kurz
0  2025-10-01  24.295
1  2025-10-02  24.240
2  2025-10-03  24.270
3  2025-10-06  24.305
4  2025-10-07  24.320
5  2025-10-08  24.370
6  2025-10-09  24.320
7  2025-10-10  24.300
8  2025-10-13  24.355
9  2025-10-14  24.320
10 2025-10-15  24.280
11 2025-10-16  24.290
12 2025-10-17  24.315
13 2025-10-20  24.295
14 2025-10-21  24.315
15 2025-10-22  24.315
16 2025-10-23  24.305
17 2025-10-24  24.340
18 2025-10-27  24.325

## Čtení tabulek přímo z webové stránky

In [89]:
csu_url = 'https://csu.gov.cz/mira_inflace'
tables = pandas.read_html(csu_url,index_col=[0], decimal=',', thousands=' ')
print("Přečteno", len(tables), 'tabulek ...', type(tables))

Přečteno 4 tabulek ... <class 'list'>


In [90]:
tables[0].info()
tables[0]

<class 'pandas.core.frame.DataFrame'>
Index: 26 entries, 2000 to 2025
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   (Měsíc, 1)   26 non-null     float64
 1   (Měsíc, 2)   26 non-null     float64
 2   (Měsíc, 3)   26 non-null     float64
 3   (Měsíc, 4)   26 non-null     float64
 4   (Měsíc, 5)   26 non-null     float64
 5   (Měsíc, 6)   26 non-null     float64
 6   (Měsíc, 7)   26 non-null     float64
 7   (Měsíc, 8)   26 non-null     float64
 8   (Měsíc, 9)   26 non-null     float64
 9   (Měsíc, 10)  25 non-null     float64
 10  (Měsíc, 11)  25 non-null     float64
 11  (Měsíc, 12)  25 non-null     float64
dtypes: float64(12)
memory usage: 2.6 KB


Rok  Měsíc                                                                  
Rok      1     2     3     4     5     6     7     8     9    10    11    12
2000   2.1   2.2   2.3   2.4   2.5   2.6   2.9   3.1   3.3   3.6   3.8   3.9
2001   4.0   4.0   4.0   4.1   4.2   4.3   4.5   4.6   4.7   4.7   4.7   4.7
2002   4.6   4.6   4.6   4.5   4.3   3.9   3.5   3.1   2.7   2.4   2.1   1.8
2003   1.5   1.1   0.8   0.5   0.3   0.2   0.2   0.1   0.0   0.0   0.1   0.1
2004   0.3   0.5   0.8   1.0   1.2   1.4   1.7   2.0   2.2   2.5   2.7   2.8
2005   2.8   2.7   2.6   2.6   2.5   2.4   2.2   2.1   2.0   2.0   1.9   1.9
2006   2.0   2.1   2.2   2.3   2.4   2.5   2.6   2.7   2.8   2.7   2.6   2.5
2007   2.4   2.3   2.2   2.2   2.1   2.1   2.1   2.0   2.0   2.2   2.5   2.8
2008   3.4   3.9   4.3   4.7   5.0   5.4   5.8   6.1   6.4   6.6   6.5   6.3
2009   5.9   5.4   5.0   4.6   4.1   3.7   3.1   2.6   2.1   1.6   1.3   1.0
2010   0.9   0.8   0.7   0.6   0.6   0.6   0.8   0.9   1.1   1.2   1.4   1.5
2011   1.6   1.7   1.7   1.8   1.8   1.9   1.9   1.9   1.8   1.9   1.9   1.9
2012   2.1   2.2   2.4   2.6   2.7   2.8   2.9   3.1   3.2   3.3   3.3   3.3
2013   3.2   3.0   2.8   2.7   2.5   2.3   2.2   2.0   1.8   1.6   1.5   1.4
2014   1.3   1.1   1.0   0.9   0.8   0.7   0.6   0.5   0.5   0.5   0.5   0.4
2015   0.3   0.3   0.3   0.4   0.4   0.5   0.5   0.4   0.4   0.4   0.3   0.3
2016   0.4   0.4   0.4   0.4   0.4   0.3   0.3   0.3   0.3   0.4   0.5   0.7
2017   0.8   1.0   1.2   1.3   1.5   1.7   1.8   2.0   2.2   2.3   2.4   2.5
2018   2.4   2.4   2.3   2.3   2.3   2.3   2.3   2.3   2.3   2.2   2.2   2.1
2019   2.2   2.3   2.4   2.4   2.5   2.5   2.6   2.6   2.6   2.7   2.7   2.8
2020   2.9   3.0   3.1   3.1   3.1   3.1   3.2   3.2   3.3   3.3   3.2   3.2
2021   3.0   2.9   2.8   2.8   2.8   2.8   2.8   2.8   3.0   3.2   3.5   3.8
2022   4.5   5.2   6.1   7.0   8.1   9.4  10.6  11.7  12.7  13.5  14.4  15.1
2023  15.7  16.2  16.4  16.2  15.8  15.1  14.3  13.6  12.7  12.1  11.4  10.7
2024   9.4   8.2   7.1   6.3   5.6   4.9   4.4   3.9   3.5   3.1   2.7   2.4
2025   2.5   2.5   2.6   2.5   2.5   2.6   2.6   2.6   2.6   NaN   NaN   NaN

## Čtení více `DataFrame` do `dict`

In [91]:
MENY = ('EUR','USD','GBP')
KURZY_URL = "https://www.cnb.cz/cs/financni-trhy/devizovy-trh/kurzy-devizoveho-trhu/kurzy-devizoveho-trhu/vybrane.txt?od=01.10.2025&do=31.10.2025&mena={mena}&format=txt"

kurzy = {

}

for mena in MENY:
    url = KURZY_URL.format(mena=mena)
    print(f"cteme kurzy '[mena]' z:", url)

    df = pandas.read_table(url,**cnb_cfg)
    kurzy[mena] = df

print("stahli jsme si kurzy:",kurzy.keys())


cteme kurzy '[mena]' z: https://www.cnb.cz/cs/financni-trhy/devizovy-trh/kurzy-devizoveho-trhu/kurzy-devizoveho-trhu/vybrane.txt?od=01.10.2025&do=31.10.2025&mena=EUR&format=txt
cteme kurzy '[mena]' z: https://www.cnb.cz/cs/financni-trhy/devizovy-trh/kurzy-devizoveho-trhu/kurzy-devizoveho-trhu/vybrane.txt?od=01.10.2025&do=31.10.2025&mena=USD&format=txt
cteme kurzy '[mena]' z: https://www.cnb.cz/cs/financni-trhy/devizovy-trh/kurzy-devizoveho-trhu/kurzy-devizoveho-trhu/vybrane.txt?od=01.10.2025&do=31.10.2025&mena=GBP&format=txt
stahli jsme si kurzy: dict_keys(['EUR', 'USD', 'GBP'])


In [92]:
kurzy.get('GBP')

Datum    Kurz
0  2025-10-01  27.921
1  2025-10-02  27.796
2  2025-10-03  27.811
3  2025-10-06  27.955
4  2025-10-07  27.939
5  2025-10-08  28.099
6  2025-10-09  28.008
7  2025-10-10  27.897
8  2025-10-13  28.071
9  2025-10-14  27.912
10 2025-10-15  27.896
11 2025-10-16  27.990
12 2025-10-17  27.956
13 2025-10-20  27.956
14 2025-10-21  28.019
15 2025-10-22  27.981
16 2025-10-23  27.965
17 2025-10-24  27.894
18 2025-10-27  27.891

## Uložení `DataFrame` do souboru

### CSV

In [93]:
CSV = os.path.join(DATA_FOLDER, 'working', "kurzy_{mena}.csv")

for mena, df in kurzy.items():
    df.to_csv(CSV.format(mena=mena, index=False))

### Excel

In [94]:
XLS = os.path.join(DATA_FOLDER, 'working', "kurzy.xlsx")
with pandas.ExcelWriter(XLS) as xlswr:
    for mena, df in kurzy.items():
        df.to_excel(xlswr, index=False,sheet_name=mena)

## Spojování `DataFrame`

### Spojení sloupců `DataFrame.join()`

vhodné pro přehledové tabulky /csv, excel/ a jednorázové analýzy

| Datum | EUR | USD | GBP |
|-------|------|------|------|
| 2023-12-01 | 24.35 | 22.39 | 28.3 |
| 2023-12-04 | 24.39 | 22.44 | 28.43 |
| 2023-12-05 | 24.37 | 22.52 | 28.44 |
| 2023-12-06 | 24.36 | 22.59 | 28.46 |
| 2023-12-07 | 24.36 | 22.62 | 28.4 |
| 2023-12-08 | 24.36 | 22.6 | 28.42 |
| 2023-12-11 | 24.38 | 22.66 | 28.49 |
| 2023-12-12 | 24.43 | 22.6 | 28.44 |
| 2023-12-13 | 24.48 | 22.7 | 28.42 |
| 2023-12-14 | 24.41 | 22.38 | 28.4 |
| 2023-12-15 | 24.48 | 22.36 | 28.52 |
| 2023-12-18 | 24.55 | 22.49 | 28.46 |



In [95]:
for mena, df in kurzy.items():
    df.rename(columns={'Kurz':mena}, inplace=True)
kurzy.get('USD')

Datum     USD
0  2025-10-01  20.721
1  2025-10-02  20.618
2  2025-10-03  20.683
3  2025-10-06  20.809
4  2025-10-07  20.855
5  2025-10-08  20.945
6  2025-10-09  20.948
7  2025-10-10  20.999
8  2025-10-13  21.046
9  2025-10-14  21.039
10 2025-10-15  20.890
11 2025-10-16  20.847
12 2025-10-17  20.819
13 2025-10-20  20.839
14 2025-10-21  20.948
15 2025-10-22  20.978
16 2025-10-23  20.968
17 2025-10-24  20.959
18 2025-10-27  20.897

In [96]:
kurzy_pivot = kurzy.get(MENY[0]).set_index('Datum').join([kurzy.get(mena).set_index('Datum') for mena in MENY[1:]])
kurzy_pivot

EUR     USD     GBP
Datum                             
2025-10-01  24.295  20.721  27.921
2025-10-02  24.240  20.618  27.796
2025-10-03  24.270  20.683  27.811
2025-10-06  24.305  20.809  27.955
2025-10-07  24.320  20.855  27.939
2025-10-08  24.370  20.945  28.099
2025-10-09  24.320  20.948  28.008
2025-10-10  24.300  20.999  27.897
2025-10-13  24.355  21.046  28.071
2025-10-14  24.320  21.039  27.912
2025-10-15  24.280  20.890  27.896
2025-10-16  24.290  20.847  27.990
2025-10-17  24.315  20.819  27.956
2025-10-20  24.295  20.839  27.956
2025-10-21  24.315  20.948  28.019
2025-10-22  24.315  20.978  27.981
2025-10-23  24.305  20.968  27.965
2025-10-24  24.340  20.959  27.894
2025-10-27  24.325  20.897  27.891

In [97]:
pandas.melt(kurzy_pivot, ignore_index=False)

variable   value
Datum                      
2025-10-01      EUR  24.295
2025-10-02      EUR  24.240
2025-10-03      EUR  24.270
2025-10-06      EUR  24.305
2025-10-07      EUR  24.320
2025-10-08      EUR  24.370
2025-10-09      EUR  24.320
2025-10-10      EUR  24.300
2025-10-13      EUR  24.355
2025-10-14      EUR  24.320
2025-10-15      EUR  24.280
2025-10-16      EUR  24.290
2025-10-17      EUR  24.315
2025-10-20      EUR  24.295
2025-10-21      EUR  24.315
2025-10-22      EUR  24.315
2025-10-23      EUR  24.305
2025-10-24      EUR  24.340
2025-10-27      EUR  24.325
2025-10-01      USD  20.721
2025-10-02      USD  20.618
2025-10-03      USD  20.683
2025-10-06      USD  20.809
2025-10-07      USD  20.855
2025-10-08      USD  20.945
2025-10-09      USD  20.948
2025-10-10      USD  20.999
2025-10-13      USD  21.046
2025-10-14      USD  21.039
2025-10-15      USD  20.890
2025-10-16      USD  20.847
2025-10-17      USD  20.819
2025-10-20      USD  20.839
2025-10-21      USD  20.948
2025-10-22      USD  20.978
2025-10-23      USD  20.968
2025-10-24      USD  20.959
2025-10-27      USD  20.897
2025-10-01      GBP  27.921
2025-10-02      GBP  27.796
2025-10-03      GBP  27.811
2025-10-06      GBP  27.955
2025-10-07      GBP  27.939
2025-10-08      GBP  28.099
2025-10-09      GBP  28.008
2025-10-10      GBP  27.897
2025-10-13      GBP  28.071
2025-10-14      GBP  27.912
2025-10-15      GBP  27.896
2025-10-16      GBP  27.990
2025-10-17      GBP  27.956
2025-10-20      GBP  27.956
2025-10-21      GBP  28.019
2025-10-22      GBP  27.981
2025-10-23      GBP  27.965
2025-10-24      GBP  27.894
2025-10-27      GBP  27.891

### Spojení řádků `pandas.concat()`

vhodné k ukládání do DB pro opakované analýzy a grafické prezentace


| Datum | Kurz | Měna |
|-------|------|------|
| 2023-12-01 | 24.35 | EUR |
| 2023-12-01 | 22.39 | USD |
| 2023-12-01 | 28.3 | GBP |
| 2023-12-01 | 25.55 | CHF |
| 2023-12-01 | 2.08 | NOK |
| 2023-12-04 | 24.39 | EUR |
| 2023-12-04 | 22.44 | USD |
| 2023-12-04 | 28.43 | GBP |
| 2023-12-04 | 25.73 | CHF |
| 2023-12-04 | 2.09 | NOK |


In [98]:
kurzy_xls = pandas.read_excel(XLS, sheet_name=None)
kurzy_xls

{'EUR':         Datum    Kurz
 0  2025-10-01  24.295
 1  2025-10-02  24.240
 2  2025-10-03  24.270
 3  2025-10-06  24.305
 4  2025-10-07  24.320
 5  2025-10-08  24.370
 6  2025-10-09  24.320
 7  2025-10-10  24.300
 8  2025-10-13  24.355
 9  2025-10-14  24.320
 10 2025-10-15  24.280
 11 2025-10-16  24.290
 12 2025-10-17  24.315
 13 2025-10-20  24.295
 14 2025-10-21  24.315
 15 2025-10-22  24.315
 16 2025-10-23  24.305
 17 2025-10-24  24.340
 18 2025-10-27  24.325,
 'USD':         Datum    Kurz
 0  2025-10-01  20.721
 1  2025-10-02  20.618
 2  2025-10-03  20.683
 3  2025-10-06  20.809
 4  2025-10-07  20.855
 5  2025-10-08  20.945
 6  2025-10-09  20.948
 7  2025-10-10  20.999
 8  2025-10-13  21.046
 9  2025-10-14  21.039
 10 2025-10-15  20.890
 11 2025-10-16  20.847
 12 2025-10-17  20.819
 13 2025-10-20  20.839
 14 2025-10-21  20.948
 15 2025-10-22  20.978
 16 2025-10-23  20.968
 17 2025-10-24  20.959
 18 2025-10-27  20.897,
 'GBP':         Datum    Kurz
 0  2025-10-01  27.921
 1  2025-10

In [99]:
for mena, df in kurzy_xls.items():
    df['Mena'] = mena
    df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19 entries, 0 to 18
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   Datum   19 non-null     datetime64[ns]
 1   Kurz    19 non-null     float64       
 2   Mena    19 non-null     object        
dtypes: datetime64[ns](1), float64(1), object(1)
memory usage: 588.0+ bytes
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19 entries, 0 to 18
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   Datum   19 non-null     datetime64[ns]
 1   Kurz    19 non-null     float64       
 2   Mena    19 non-null     object        
dtypes: datetime64[ns](1), float64(1), object(1)
memory usage: 588.0+ bytes
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19 entries, 0 to 18
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   Datum   1

In [100]:
kurzy_ts = pandas.concat(kurzy_xls.values(), ignore_index=True)
kurzy_ts.to_csv(os.path.join(DATA_FOLDER,'working','kurzy_ts.csv'), index=False)

## Jak na SQL?

In [101]:
DBURI = "postgresql://neondb_owner:npg_MRPdz0J5KfNg@ep-plain-rice-agzd949r-pooler.c-2.eu-central-1.aws.neon.tech/DAPYTPCS?sslmode=require&channel_binding=require"
SQLDB = sql.create_engine(DBURI)

### Uložení `DataFrame` do SQL databaze

In [107]:
kurzy_ts.set_index("Datum").to_sql("kurzy",SQLDB,if_exists="replace")

57

### Čtení `DataFrame` z SQL databáze

In [110]:
kurzy_sql = pandas.read_sql_table("kurzy",SQLDB,index_col="Datum")

In [111]:
kurzy_sql.info()
kurzy_sql

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 57 entries, 2025-10-01 to 2025-10-27
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Kurz    57 non-null     float64
 1   Mena    57 non-null     object 
dtypes: float64(1), object(1)
memory usage: 1.3+ KB


Kurz Mena
Datum                  
2025-10-01  24.295  EUR
2025-10-02  24.240  EUR
2025-10-03  24.270  EUR
2025-10-06  24.305  EUR
2025-10-07  24.320  EUR
2025-10-08  24.370  EUR
2025-10-09  24.320  EUR
2025-10-10  24.300  EUR
2025-10-13  24.355  EUR
2025-10-14  24.320  EUR
2025-10-15  24.280  EUR
2025-10-16  24.290  EUR
2025-10-17  24.315  EUR
2025-10-20  24.295  EUR
2025-10-21  24.315  EUR
2025-10-22  24.315  EUR
2025-10-23  24.305  EUR
2025-10-24  24.340  EUR
2025-10-27  24.325  EUR
2025-10-01  20.721  USD
2025-10-02  20.618  USD
2025-10-03  20.683  USD
2025-10-06  20.809  USD
2025-10-07  20.855  USD
2025-10-08  20.945  USD
2025-10-09  20.948  USD
2025-10-10  20.999  USD
2025-10-13  21.046  USD
2025-10-14  21.039  USD
2025-10-15  20.890  USD
2025-10-16  20.847  USD
2025-10-17  20.819  USD
2025-10-20  20.839  USD
2025-10-21  20.948  USD
2025-10-22  20.978  USD
2025-10-23  20.968  USD
2025-10-24  20.959  USD
2025-10-27  20.897  USD
2025-10-01  27.921  GBP
2025-10-02  27.796  GBP
2025-10-03  27.811  GBP
2025-10-06  27.955  GBP
2025-10-07  27.939  GBP
2025-10-08  28.099  GBP
2025-10-09  28.008  GBP
2025-10-10  27.897  GBP
2025-10-13  28.071  GBP
2025-10-14  27.912  GBP
2025-10-15  27.896  GBP
2025-10-16  27.990  GBP
2025-10-17  27.956  GBP
2025-10-20  27.956  GBP
2025-10-21  28.019  GBP
2025-10-22  27.981  GBP
2025-10-23  27.965  GBP
2025-10-24  27.894  GBP
2025-10-27  27.891  GBP

## Vytvoření `DataFrame` z `dict` 

In [113]:
KEY_NAME ='jmeno'
KEY_PARTNER = 'partner'
KEY_CHILDREN = 'deti'
KEY_EXTRAS = 'zajimavost'

anna = {KEY_NAME: 'Anna Nova'}
boris = {KEY_NAME: 'Boris Novy'}
cecilie = {KEY_NAME: 'Cecilie Nova'}
david = {KEY_NAME: 'David Stary'}
filip = {KEY_NAME: 'Filip stary'}
gustav = {KEY_NAME: 'Gustav Stary'}

rodina = [anna, boris, cecilie, david, filip, gustav]

filip [KEY_EXTRAS] = "dvojcata"
gustav [KEY_EXTRAS] = "dvojcata"
anna [KEY_EXTRAS] = "zahradnice"
boris [KEY_EXTRAS] = "sportovec"
david [KEY_EXTRAS] = "cestovatel"
cecilie [KEY_EXTRAS] = "cukrářka"

rodina_df=pandas.DataFrame(rodina)
rodina_df.info()
rodina_df

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6 entries, 0 to 5
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   jmeno       6 non-null      object
 1   zajimavost  6 non-null      object
dtypes: object(2)
memory usage: 228.0+ bytes


jmeno  zajimavost
0     Anna Nova  zahradnice
1    Boris Novy   sportovec
2  Cecilie Nova    cukrářka
3   David Stary  cestovatel
4   Filip stary    dvojcata
5  Gustav Stary    dvojcata

Světem internetu se toulá robot, který každou minutu vysílá informace o svém aktuálním stavu:
- svůj čas: `cas_robota` index 
- textovou zprávu: `zprava`
- GPS pozici: `lati`,`longi`
- hodnoty hodu šesti kostkami: `kostka1`, `kostka2`, `kostka3`, `kostka4`, `kostka5`, `kostka6`
- cenu, stav počítadla a senzoru: `cena`,`pocitadlo`,`senzor` 

1. Připravíme si startovací zprávu

In [73]:
robot = dict(
    zprava="Rip: start robotovy cesty",
    lati=50.30688756,
    longi=14.2896057,
    kostka1=random.randrange(1,7),
    kostka2=random.randrange(1,7),
    kostka3=random.randrange(1,7),
    kostka4=random.randrange(1,7),
    kostka5=random.randrange(1,7),
    kostka6=random.randrange(1,7),
    cena=1000*random.random()+2000,
    pocitadlo=random.randrange(500,1500),
    senzor=900*random.random()*200,
)

robot

{'zprava': 'Rip: start robotovy cesty',
 'lati': 50.30688756,
 'longi': 14.2896057,
 'kostka1': 5,
 'kostka2': 5,
 'kostka3': 6,
 'kostka4': 6,
 'kostka5': 3,
 'kostka6': 5,
 'cena': 2142.476984342657,
 'pocitadlo': 1046,
 'senzor': 42728.71766652427}

2. Vytvoříme `DataFrame`

In [76]:
robot_df = pandas.DataFrame(robot,[pandas.Timestamp.now()])
robot_df.index.rename("cas_robota",inplace=True)
robot_df

zprava       lati      longi  \
cas_robota                                                                    
2025-10-23 21:08:51.427796  Rip: start robotovy cesty  50.306888  14.289606   

                            kostka1  kostka2  kostka3  kostka4  kostka5  \
cas_robota                                                                
2025-10-23 21:08:51.427796        5        5        6        6        3   

                            kostka6         cena  pocitadlo        senzor  
cas_robota                                                                 
2025-10-23 21:08:51.427796        5  2142.476984       1046  42728.717667

3. Uložíme do databáze

In [ ]:
#robot_df.to_sql('cesta_robota',SQLDB)

1

In [ ]:
dbtypes = dict(
    zprava=sql.types.Text,
    lati=sql.types.Double,
    longi=sql.types.Double,
    kostka1=sql.types.SmallInteger,
    kostka2=sql.types.SmallInteger,
    kostka3=sql.types.SmallInteger,
    kostka4=sql.types.SmallInteger,
    kostka5=sql.types.SmallInteger,
    kostka6=sql.types.SmallInteger,
    cena=sql.types.Numeric(10,3),
    pocitadlo=sql.types.Integer,
    senzor=sql.types.Double,
)
#robot_df.to_sql('cesta_robota',SQLDB,dtype=dbtypes,if_exists='replace')

1

4. otestujeme přístup `studenta` a `robota` k nové tabulce  
_Poznámka:_ Pokud to neprojde, spusť nad svojí databází [`resource/sql/db-grants.sql`](../resource/sql/db-grants.sql)

In [ ]:
#

In [ ]:
#